### PR TITLE
Add map-in-map type checking

### LINKED ISSUE #375: Support BPF_MAP_TYPE_HASH_OF_MAPS and BPF_MAP_TYPE_ARRAY_OF_MAPS
This was tracked as part of issue #343.  Splitting into its own issue so it can have its own milestone and assignee.

### PR DESCRIPTION
**Pull request by dthaler**
This PR enforces that all inner maps must be of the same type
as the inner map template used for verification.  Other fields
might need to match too, and if so those will be updated in
a subsequent PR once it is confirmed which fields must match.

A few pieces of this PR related to map_id are prerequisites
for issue #396 which will add IDs for programs, maps, and links.

Finally, there are multiple definitions of bpf_map, since the
version used to write eBPF programs is different from what is
stored in memory (which uses map IDs) so to avoid confusion in
code and allow the compiler to do type checking to catch some
bugs, this splits ebpf_map_definition_t into two, one for
in_memory and one for in_file (meaning in an eBPF program).
This will also allow the future PR for issue 396 to be more
understandable, but also aids clarity in some parts of this PR.

Terminology for reference when reviewing this PR and future PR for 396 (these are also in comments in files but summarized here):
* map_id: a 32-bit number created by the execution context that is sort of like an fd for kernel-mode.
* map_idx (as  used in "inner_map_idx" in the verifier): For maps loaded from the maps section, this is the index in the maps section.
* original_fd: the "fd" used in an eBPF program before relocation.  For maps loaded from the maps section, this is the map index + 1.

Finally, this also adds the SEC() macro for cross-plat compat for eBPF programs.

Fixes #375 

Signed-off-by: Dave Thaler <dthaler@microsoft.com>

### CODE REVIEW

File: libs/execution_context/ebpf_maps.c [hunk 1]
```diff
@@ -264,6 +265,26 @@ _associate_program_with_prog_array_map(_In_ ebpf_core_map_t* map, _In_ const ebp
...
+    if (outer_map->ebpf_map_definition.type != BPF_MAP_TYPE_ARRAY_OF_MAPS &&
+        outer_map->ebpf_map_definition.type != BPF_MAP_TYPE_HASH_OF_MAPS) {
+        return true;
+    }
+
+    ebpf_core_map_t* template = outer_map->inner_map_template;
+    ebpf_object_t* template_object = &template->object;
+    const ebpf_map_t* value_map = (ebpf_map_t*)value_object;
+
+    // TODO: what other map attributes need to match?
```
Comment (Alan-Jowett) [inline libs/execution_context/ebpf_maps.c:280]: File an issue for this so it doesn't get lost.

File: libs/execution_context/ebpf_maps.c [hunk 2]
```diff
@@ -264,6 +265,26 @@ _associate_program_with_prog_array_map(_In_ ebpf_core_map_t* map, _In_ const ebp
     return result;
 }
 
+static bool // Returns true if ok, false if not.
+_check_value_type(_In_ const ebpf_core_map_t* outer_map, _In_ const ebpf_object_t* value_object)
```
Comment (Alan-Jowett) [inline libs/execution_context/ebpf_maps.c:269]: Maybe add an explanation that this function releases a reference on inner_map_template?

File: libs/execution_context/ebpf_maps.c [hunk 3]
```diff
@@ -875,6 +908,13 @@ ebpf_map_create(
         goto Exit;
     }
 
+    if (inner_map_handle != (ebpf_handle_t)-1) {
+        // Convert value handle to an object pointer.
+        result = ebpf_reference_object_by_handle(inner_map_handle, EBPF_OBJECT_MAP, &inner_map_template_object);
```
Comment (Alan-Jowett) [inline libs/execution_context/ebpf_maps.c:913]: The pattern so far has been to convert all handle -> object references in ebpf_core.c.

**Review by Alan-Jowett — COMMENTED** [review]
Some questions about how we are using the handles, but the rest looks good to me.

File: libs/execution_context/ebpf_maps.c (hunk 2 above)
Comment (dthaler) [reply libs/execution_context/ebpf_maps.c:269]: It shouldn't, I see a bug in the caller that will leak a reference for PROG_ARRAY, let me clean this up.

File: libs/execution_context/ebpf_maps.c (hunk 1 above)
Comment (dthaler) [reply libs/execution_context/ebpf_maps.c]: Done

File: libs/execution_context/ebpf_maps.c (hunk 2 above)
Comment (dthaler) [reply libs/execution_context/ebpf_maps.c:269]: Actually it shouldn't, there was a bug in my code which I have now fixed.

File: libs/execution_context/ebpf_maps.c (hunk 3 above)
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_maps.c:913]: Based on an offline discussion, this is an intentional move away from having all handle -> object conversion in ebpf_core.c and towards having the map specific logic in ebpf_map.c.

**Review by Alan-Jowett — APPROVED** [review]
LGTM. Mostly focuses on the EC part, but the API/lib part looks ok too (I am not as familiar with it though).

File: libs/api/windows_platform.cpp [hunk 4]
```diff
@@ -13,83 +13,60 @@
...
-    uint32_t value_size,
-    uint32_t max_entries,
-    uint32_t inner_map_fd,
-    size_t section_offset,
-    ebpf_verifier_options_t options)
-{
-    int fd;
-    if (options.mock_map_fds) {
-        EbpfMapType type = get_map_type_windows(map_type);
-        fd = create_map_crab(type, key_size, value_size, max_entries, options);
```
Comment (saxena-anurag) [inline libs/api/windows_platform.cpp:39]: When verifier calls the Windows callback function `parse_maps_section_windows()` and passes `mock_map_fds` option in the verifier options, this function used to call `create_map_crab()` which used to then populate some TLS in the verifier code.
Do we not need that flow now?

File: libs/api/windows_platform.cpp [hunk 5]
```diff
@@ -13,83 +13,60 @@
...
-                s.value_size,
-                s.max_entries,
-                s.inner_map_idx,
-                (i * sizeof(ebpf_map_definition_t)),
-                options),
+        uint32_t section_offset = (i * sizeof(ebpf_map_definition_in_file_t));
+
+        // The map file descriptors that appear in eBPF bytecode start at 1,
+        // in the order the maps appear in the maps section.
+        const int ORIGINAL_FD_OFFSET = 1;
```
Comment (saxena-anurag) [inline libs/api/windows_platform.cpp:38]: minor: We can move this outside the for loop.

File: libs/api/windows_platform.cpp [hunk 6]
```diff
@@ -13,83 +13,60 @@
...
+            .inner_map_fd = inner_map_original_fd});
     }
 }
 
 const ebpf_platform_t g_ebpf_platform_windows = {
     get_program_type_windows,
     get_helper_prototype_windows,
     is_helper_usable_windows,
-    sizeof(ebpf_map_definition_t),
+    sizeof(ebpf_map_definition_in_memory_t),
```
Comment (saxena-anurag) [inline libs/api/windows_platform.cpp:69]: Should this be `ebpf_map_definition_in_file_t`?

File: libs/api/windows_platform.cpp (hunk 4 above)
Comment (dthaler) [reply libs/api/windows_platform.cpp:39]: Nope, the flow is simplified so it's done differently.  The old flow was pretty convoluted.

File: libs/api/windows_platform.cpp (hunk 6 above)
Comment (dthaler) [reply libs/api/windows_platform.cpp:69]: Yes thanks.  (Though at present they're the same size.)

File: libs/api/windows_platform.cpp (hunk 5 above)
Comment (dthaler) [reply libs/api/windows_platform.cpp]: Done

File: libs/api/ebpf_api.cpp [hunk 7]
```diff
@@ -1641,6 +1625,51 @@ _initialize_ebpf_object_from_elf(
...
+        if (map->inner_map == nullptr) {
+            // This map requires an inner map template, look up which one.
+            for (auto& inner_map : maps) {
+                if (inner_map->original_fd == map->inner_map_original_fd) {
+                    map->inner_map = inner_map;
+                    break;
+                }
+            }
+            if (map->inner_map == nullptr) {
+                // We can't create this map because there is no inner template.
```
Comment (saxena-anurag) [inline libs/api/ebpf_api.cpp:1653]: If inner map template is not present, then the whole program load should fail?

Comment (dthaler) [reply libs/api/ebpf_api.cpp:1653]: I don't think it fails, I think it just doesn't create the outer map.

Comment (saxena-anurag) [reply libs/api/ebpf_api.cpp:1653]: I think the behavior should be all or none. libbpf also has similar behavior. If inner map template is not found, then we should return error.

Comment (saxena-anurag) [reply libs/api/ebpf_api.cpp:1653]: The same is true for programs also. If loading of one program fails, the whole API fails.

File: tests/sample/bindmonitor.c [hunk 8]
```diff
@@ -16,15 +16,15 @@ typedef struct _process_entry
 } process_entry_t;
 
 #pragma clang section data = "maps"
-ebpf_map_definition_t process_map = {
-    .size = sizeof(ebpf_map_definition_t),
+ebpf_map_definition_in_file_t process_map = {
```
Comment (saxena-anurag) [inline tests/sample/bindmonitor.c:19]: As we have changed the name of this struct, will this cause problems with portability of ebpf programs?

File: libs/api/ebpf_api.cpp (hunk 7 above)
Comment (dthaler) [reply libs/api/ebpf_api.cpp:1654]: Updated

**Review by saxena-anurag — APPROVED** [review]

**Pull request merged**